presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -384,14 +446,64 @@ private RowExpression notCallExpression(RowExpression argument)
         {
             return new CallExpression("not", functionResolution.notFunction(), BOOLEAN, singletonList(argument));
         }
+    }
 
-        private boolean isConjunctionOrDisjunction(RowExpression expression)
+    private class ConvertNormalFormVisitor
+            implements RowExpressionVisitor<RowExpression, Form>
+    {
+        @Override
+        public RowExpression visitSpecialForm(SpecialFormExpression specialForm, Form clauseJoiner)
         {
-            if (expression instanceof SpecialFormExpression) {
-                Form form = ((SpecialFormExpression) expression).getForm();
-                return form == AND || form == OR;
+            if (!isConjunctionOrDisjunction(specialForm)) {
+                return specialForm;
+            }
+
+            // normalize arguments
+            RowExpression left = specialForm.getArguments().get(0).accept(new ConvertNormalFormVisitor(), clauseJoiner);
+            RowExpression right = specialForm.getArguments().get(1).accept(new ConvertNormalFormVisitor(), clauseJoiner);
+
+            // check if already in correct form
+            if (specialForm.getForm() == clauseJoiner) {
+                return clauseJoiner == AND ? and(left, right) : or(left, right);
             }
-            return false;
+
+            // else, we expand and distribute
+            Form termJoiner = clauseJoiner == AND ? OR : AND;
+            List<RowExpression> leftClauses = extractPredicates(clauseJoiner, left);
+            List<RowExpression> rightClauses = extractPredicates(clauseJoiner, right);
+            List<RowExpression> permutationClauses = new ArrayList<>();
+            leftClauses.forEach(leftClause -> rightClauses.forEach(rightClause -> permutationClauses.add(termJoiner == AND ? and(leftClause, rightClause) : or(leftClause, rightClause))));
```

Comment:
What about writing nested for-loop? -- My personal experience is lambda, when used in a nested way, is not easy to read 😛